elastic/beats
# [POC] Filebeat azure input based on the azure event hubs sdk

Initial issue https://github.com/elastic/beats/issues/14092
Filebeat azure input based on the azure event hubs sdk
using https://github.com/Azure/azure-event-hubs-go

Input is called: azureeventhub to match existing input naming.

Configuration options:
```
filebeat.inputs:
- type: azure-eventhub
  eventhub: "insights-operational-logs"
  consumer_group: "obstest"
  connection_string: "Endpoint=sb://....."
```
Configuration contains eventhub name, connection string to eventhub namespace and an optional consumer group.


To start back up at the spot the input stopped processing messages the message offsets should be stored. 
The azure-eventhub  Hub struct can be customized to use an persist.CheckpointPersister. By default, a Hub uses an in-memory CheckpointPersister, but accepts anything that implements the perist.CheckpointPersister interface.
The in-memory persister is not reliable, during the tests offsets were lost.
They also mention:
```
// MemoryPersister is only intended to be shared with EventProcessorHosts within the same process. This implementation
// is a toy. You should probably use the Azure Storage implementation or any other that provides durable storage for
// checkpoints.`
````
At the moment a FilePersister is used, offsets are stored in files in the data/azureeventhubpersister directory.


If users want the events to be consumed using EPH (EPH is intended to be run across multiple processes and machines while load balancing message consumers more on this here https://github.com/Azure/azure-event-hubs-go#event-processor-host, https://docs.microsoft.com/en-us/azure/event-hubs/event-hubs-event-processor-host) they can enable the option `enable_eph` and add information on the leaser/checkpointer storage. 


```
filebeat.inputs:
- type: azure-eventhub
  eventhub: "insights-operational-logs"
  consumer_group: "obstest"
  connection_string: "Endpoint=sb://....."
  enable_eph: true
  storage_account: "azuresaeh"
  storage_account_key: "....."
  storage_account_container: ""
```

Event output:
```
 {
        "_index" : "filebeat-8.0.0-2019.12.09",
        "_type" : "_doc",
        "_id" : "E2oV8G4BmZ3Iz4oUDJu6",
        "_score" : null,
        "_source" : {
          "@timestamp" : "2019-12-09T20:21:29.054Z",
          "host" : {
            "name" : "DESKTOP-RFOOE09",
            "hostname" : "DESKTOP-RFOOE09",
            "architecture" : "x86_64",
            "os" : {
              "build" : "18362.476",
              "platform" : "windows",
              "version" : "10.0",
              "family" : "windows",
              "name" : "Windows 10 Pro",
              "kernel" : "10.0.18362.476 (WinBuild.160101.0800)"
            },
            "id" : "1e50b6e1-9710-4164-a8f0-032b3c721dc3"
          },
          "agent" : {
            "hostname" : "DESKTOP-RFOOE09",
            "id" : "17cead09-57ad-4668-8a0e-b9025f8b0cb0",
            "version" : "8.0.0",
            "type" : "filebeat",
            "ephemeral_id" : "92de04b7-6096-469b-bd70-cfa059683e9d"
          },
          "message" : """{"callerIpAddress":"77.170.179.229","category":"Action","correlationId":"d5dbab34-f6fb-451a-b361-80eb37b6b3d7","durationMs":909,"identity":{"authorization":.............................."}""",
          "azure" : {
            "offset" : 4982456,
            "sequence_number" : 563,
            "enqueued_time" : "2019-12-17T13:47:36.09Z",
            "partition_id" : 0,
            "eventhub" : "insights-operational-logs",
            "consumer_group" : "obstest"
          },
          "input" : {
            "type" : "azure-eventhub"
          },
          "ecs" : {
            "version" : "1.2.0"
          }
        },
        "sort" : [
          1575922889054
        ]
      },
```



Update:

Configuration will remain the same:
```
filebeat.inputs:
- type: azure-eventhub
  eventhub: "insights-operational-logs"
  consumer_group: "obstest"
  connection_string: "Endpoint=sb://....."
```

By default, an azure Hub uses an in-memory CheckpointPersister, but accepts anything that implements the perist.CheckpointPersister interface.
The in-memory persister will be selected in this case, **users should be aware** , this is not reliable, offsets will be lost if filebeat stops.

when EPH is enabled:
```
filebeat.inputs:
- type: azure-eventhub
  eventhub: "insights-operational-logs"
  consumer_group: "obstest"
  connection_string: "Endpoint=sb://....."
  enable_eph: true
  storage_account: "azuresaeh"
  storage_account_key: "....."
  storage_account_container: ""
```
Offsets are stored in the azure storage account.
